espnet2/gan_codec/shared/quantizer/modules/core_vq.py

```diff
@@ -315,24 +465,150 @@ def forward(self, x):
+        kmeans_iters: int = 50,
+        threshold_ema_dead_code: int = 2,
+        codebook_type: str = "euclidean",
+        commitment_weight: float = 0.25,
+        codebook_weight: float = 1.0,
+    ):
+        super().__init__()
+        _codebook_dim: int = default(codebook_dim, dim)
+        requires_projection = _codebook_dim != dim
+        self.project_in = (
+            nn.Linear(dim, _codebook_dim) if requires_projection else nn.Identity()
+        )
+        self.project_out = (
+            nn.Linear(_codebook_dim, dim) if requires_projection else nn.Identity()
+        )
+
+        self.epsilon = epsilon
+        self.commitment_weight = commitment_weight
+        self.codebook_weight = codebook_weight
+
+        if codebook_type == "euclidean":
+            self._codebook = EuclideanCodebook(
+                dim=_codebook_dim,
+                codebook_size=codebook_size,
+                kmeans_init=kmeans_init,
+                kmeans_iters=kmeans_iters,
+                decay=decay,
+                epsilon=epsilon,
+                threshold_ema_dead_code=threshold_ema_dead_code,
+            )
+        elif codebook_type == "cosine":
+            self._codebook = CosineSimilarityCodebook(
+                dim=_codebook_dim,
+                codebook_size=codebook_size,
+                kmeans_init=kmeans_init,
+                kmeans_iters=kmeans_iters,
+                decay=decay,
+                epsilon=epsilon,
+                threshold_ema_dead_code=threshold_ema_dead_code,
+            )
+        else:
+            raise ValueError("Unsupported codebook type: {}".format(codebook_type))
+        self.codebook_size = codebook_size
+
+    @property
+    def codebook(self):
+        return self._codebook.embed
+
+    def encode(self, x):
+        x = rearrange(x, "b d n -> b n d")
+        x = self.project_in(x)
+        embed_in = self._codebook.encode(x)
+        return embed_in
+
+    def decode(self, embed_ind):
+        quantize = self._codebook.decode(embed_ind)
+        quantize = self.project_out(quantize)
+        quantize = rearrange(quantize, "b n d -> b d n")
+        return quantize
+
+    def forward(self, x, mask):
+        device = x.device
+        x = rearrange(x, "b d n -> b n d")
+        x_e = self.project_in(x)
+        quantize, embed_ind = self._codebook(x_e)
+
+        if self.training:
+            quantize = x_e + (quantize - x_e).detach()
+
+        loss = torch.tensor([0.0], device=device, requires_grad=self.training)
+
+        if self.training:
+            if self.commitment_weight > 0:
+                commit_loss = F.mse_loss(quantize.detach(), x_e, reduction="none").mean(
+                    [1, 2]
+                )
+                loss = loss + (commit_loss * mask).mean() * self.commitment_weight
+            if self.codebook_weight > 0:
+                codebook_loss = F.mse_loss(
+                    quantize, x_e.detach(), reduction="none"
+                ).mean([1, 2])
```

Comment:
I think this implementation is the same as our quantization_loss, you may consider to merge the two